contract.sol

```diff
@@ -0,0 +1,152 @@
+
+    // Mapping to track each user's stakes based on the reward token they selected
+    mapping(address => mapping(address => Stake)) public stakes;
+    mapping(address => bool) public isRewardToken;  // Tracks if a token is a valid reward token
+
+    // Events for logging key actions in the contract
+    event Staked(address indexed user, uint256 amount, address indexed rewardToken);
+    event Withdrawn(address indexed user, uint256 amount, uint256 penalty, address indexed rewardToken);
+    event RewardClaimed(address indexed user, uint256 rewardAmount, address indexed rewardToken);
+    event PenaltiesWithdrawn(address indexed owner, uint256 amount);
+
+    // Custom errors for more efficient error handling
+    error InvalidAmount();  // Thrown when staking amount is 0 or negative
+    error InvalidRewardToken();  // Thrown when an invalid reward token is specified
+    error WithdrawAmountExceedsStake();  // Thrown when trying to withdraw more than staked
+    error NoStakeFound();  // Thrown when trying to claim reward without a stake
+    error CannotClaimRewardYet();  // Thrown when attempting to claim reward too early
+    error NoPenaltiesToWithdraw();  // Thrown when no penalties are available for withdrawal by the owner
+
+    // Initializer function to replace constructor for upgradeable contracts
+    function initialize(address _stakingToken, address[] memory _rewardTokens) public initializer {
+        __Pausable_init();  // Initialize pausable functionality
+        stakingToken = IERC20(_stakingToken);  // Set the staking token
+        minStakingPeriod = 30;  // Set the minimum staking period
+        earlyWithdrawalPenalty = 10 * 1e18;  // Set the early withdrawal penalty percentage, scaled to 1e18
+        rewardRate = 1 * 1e18;  // Set the reward rate, scaled to 1e18
+
+        // Add reward tokens and mark them as valid
+        for (uint256 i = 0; i < _rewardTokens.length; i++) {
+            rewardTokens.push(_rewardTokens[i]);
+            isRewardToken[_rewardTokens[i]] = true;
+        }
+    }
+
+    // Function to stake tokens
+    function stake(uint256 amount, address rewardToken) external nonReentrant whenNotPaused {
+        if (amount <= 0) revert InvalidAmount();  // Ensure amount is positive
+        if (!isRewardToken[rewardToken]) revert InvalidRewardToken();  // Ensure reward token is valid
+
+        stakingToken.transferFrom(msg.sender, address(this), amount);  // Transfer staked tokens to the contract
+
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        userStake.amount += amount;  // Update the staked amount
+        userStake.startTime = block.timestamp;  // Record the start time of the stake
+
+        emit Staked(msg.sender, amount, rewardToken);  // Emit a Staked event
+    }
+
+    // Function to withdraw staked tokens
+    function withdraw(uint256 amount, address rewardToken) external nonReentrant whenNotPaused {
```

Comment:
For the parameteric inputs, follow the underscored convention,
eg. use _amount instead of amount